File: src/devices/bin/driver_manager/runner.cc
// Copyright 2023 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/devices/bin/driver_manager/runner.h"

#include <fidl/fuchsia.component/cpp/common_types_format.h>
#include <lib/syslog/cpp/macros.h>
#include <zircon/processargs.h>

#include "src/devices/lib/log/log.h"

namespace {

namespace fprocess = fuchsia_process;
namespace frunner = fuchsia_component_runner;
namespace fcomponent = fuchsia_component;
namespace fdecl = fuchsia_component_decl;

constexpr uint32_t kTokenId = PA_HND(PA_USER0, 0);

zx::result<zx_koid_t> GetKoid(zx_handle_t handle) {
  zx_info_handle_basic_t info{};
  if (zx_status_t status =
          zx_object_get_info(handle, ZX_INFO_HANDLE_BASIC, &info, sizeof(info), nullptr, nullptr);
      status != ZX_OK) {
    return zx::error(status);
  }
  return zx::ok(info.koid);
}

}  // namespace

namespace driver_manager {

zx::result<> Runner::Publish(component::OutgoingDirectory& outgoing) {
  return outgoing.AddUnmanagedProtocol<frunner::ComponentRunner>(
      bindings_.CreateHandler(this, dispatcher_, fidl::kIgnoreBindingClosure));
}

void Runner::CreateDriverComponent(const std::shared_ptr<ComponentOwner>& owner,
                                   fidl::ServerEnd<fcomponent::Controller> controller_request,
                                   std::string_view moniker, std::string_view url,
                                   std::string_view collection_name,
                                   const std::vector<NodeOffer>& offers) {
  fidl::Arena arena;
  auto child_decl = fdecl::wire::Child::Builder(arena)
                        .name(fidl::StringView::FromExternal(moniker))
                        .url(fidl::StringView::FromExternal(url))
                        .startup(fdecl::wire::StartupMode::kLazy)
                        .Build();

  auto child_args_builder = fcomponent::wire::CreateChildArgs::Builder(arena);

  if (controller_request.is_valid()) {
    child_args_builder.controller(std::move(controller_request));
  }

  auto offers_dictionary = owner->TakeDictionary();

  size_t offers_count;
  if (!owner->SkipInjectedOffers()) {
    offers_count = offers.size() + offer_injector_.ExtraOffersCount();
  } else {
    offers_count = offers.size();
  }
  fidl::VectorView<fdecl::wire::Offer> dynamic_offers(arena, offers_count);
  if (!offers.empty()) {
    for (size_t i = 0; i < offers.size(); i++) {
      const NodeOffer& offer = offers[i];
      switch (offer.transport) {
        case OfferTransport::DriverTransport:
          dynamic_offers[i] = fidl::ToWire(arena, ToFidl(offer).driver_transport().value());
          break;
        case OfferTransport::ZirconTransport:
          dynamic_offers[i] = fidl::ToWire(arena, ToFidl(offer).zircon_transport().value());
          break;
      }
    }
  }
  if (!owner->SkipInjectedOffers()) {
    offer_injector_.Inject(arena, dynamic_offers, offers.size());
  }

  child_args_builder.dynamic_offers(dynamic_offers);

  if (offers_dictionary) {
    child_args_builder.dictionary(fidl::ToWire(arena, std::move(offers_dictionary.value())));
  }

  std::string child_moniker(moniker);

  auto create_callback =
      [this,
       child_moniker](fidl::WireUnownedResult<fcomponent::Realm::CreateChild>& result) mutable {
        bool is_error = false;
        if (!result.ok()) {
          fdf_log::error("Failed to create child '{}': {}", child_moniker, result.error());
          is_error = true;
        } else if (result.value().is_error()) {
          fdf_log::error("Failed to create child '{}': {}", child_moniker,
                         result.value().error_value());
          is_error = true;
        }
        if (is_error) {
          zx::result result = CallCallback(child_moniker, zx::error(ZX_ERR_INTERNAL));
          if (result.is_error()) {
            fdf_log::error("Failed to find driver request for '{}': {}", child_moniker, result);
          }

          return;
        }

        StartDriverComponent(child_moniker);
      };
  realm_
      ->CreateChild(
          fdecl::wire::CollectionRef{
              .name = fidl::StringView::FromExternal(collection_name),
          },
          child_decl, child_args_builder.Build())
      .Then(std::move(create_callback));

  moniker_to_owner_[child_moniker] = owner;
}

void Runner::StartDriverComponent(const std::string& moniker) {
  auto it = moniker_to_owner_.find(moniker);
  if (it == moniker_to_owner_.end()) {
    return;
  }

  std::shared_ptr owner = it->second.lock();
  if (!owner) {
    return;
  }

  // When we start a driver, we associate an unforgeable token (the KOID of a
  // zx::event) with the start request, through the use of the numbered_handles
  // field. We do this so:
  //  1. We can securely validate the origin of the request
  //  2. We avoid collisions that can occur when relying on the package URL
  //  3. We avoid relying on the resolved URL matching the package URL
  zx::event token;
  zx_status_t status = zx::event::create(0, &token);
  if (status != ZX_OK) {
    owner->OnComponentStarted(bootup_tracker_, std::string(moniker), zx::error(status));
    return;
  }

  zx::result koid = GetKoid(token.get());
  if (koid.is_error()) {
    owner->OnComponentStarted(bootup_tracker_, std::string(moniker), koid.take_error());
    return;
  }

  start_requests_.emplace(koid.value(), moniker);

  fprocess::wire::HandleInfo handle_info = {
      .handle = std::move(token),
      .id = kTokenId,
  };

  owner->RequestStartComponent(std::move(handle_info), moniker, bootup_tracker_);
}

void Runner::Start(StartRequestView request, StartCompleter::Sync& completer) {
  std::string url = std::string(request->start_info.resolved_url().get());

  // We use the numbered handle, if it exists, to locate the moniker of the node we are starting
  // a driver for. This will be the case for starts that we have issued ourself through
  // |Runner::StartDriverComponent|.
  auto& handles = request->start_info.numbered_handles();
  if (handles.size() == 1 && handles[0].handle && handles[0].id == kTokenId) {
    zx::result koid = GetKoid(handles[0].handle.get());
    if (koid.is_error()) {
      completer.Close(ZX_ERR_INVALID_ARGS);
      return;
    }

    auto it = start_requests_.find(koid.value());
    if (it == start_requests_.end()) {
      completer.Close(ZX_ERR_NOT_FOUND);
      return;
    }

    std::string moniker = it->second;
    start_requests_.erase(it);

    zx::result cb_result =
        CallCallback(moniker, zx::ok(StartedComponent{
                                  .info = fidl::ToNatural(request->start_info),
                                  .component_controller = std::move(request->controller),
                              }));
    if (cb_result.is_error()) {
      fdf_log::error("Failed to start driver '{}', unknown request for driver {}", url, moniker);
      completer.Close(ZX_ERR_UNAVAILABLE);
    }

    return;
  }

  // Otherwise we need to locate it using the component framework's introspection.
  // This will happen if the component framework issues a start on the component manually, which
  // can happen for various reasons, like an ffx component reload being issued.
  zx::event token;
  zx_status_t status =
      request->start_info.component_instance().duplicate(ZX_RIGHT_SAME_RIGHTS, &token);
  if (status != ZX_OK) {
    fdf_log::error("Failed to clone component_instance token.");
    completer.Close(ZX_ERR_INTERNAL);
    return;
  }
  introspector_->GetMoniker(std::move(token))
      .Then([this, completer = completer.ToAsync(),
             start_info = fidl::ToNatural(request->start_info),
             controller = std::move(request->controller),
             url](fidl::WireUnownedResult<fcomponent::Introspector::GetMoniker>& result) mutable {
        if (!result.ok()) {
          fdf_log::error("Failed to GetMoniker. {}", result.FormatDescription());
          completer.Close(ZX_ERR_INTERNAL);
          return;
        }

        if (result.value().is_error()) {
          fdf_log::error("Failed to GetMoniker. {}", result.value().error_value());
          completer.Close(ZX_ERR_INTERNAL);
          return;
        }

        std::string moniker(result.value()->moniker.get());
        size_t split_point = moniker.find(':');
        if (split_point <= 0) {
          fdf_log::error("moniker does not contain collection");
          completer.Close(ZX_ERR_INVALID_ARGS);
          return;
        }

        moniker = moniker.substr(split_point + 1);
        zx::result cb_result =
            CallCallback(moniker, zx::ok(StartedComponent{
                                      .info = std::move(start_info),
                                      .component_controller = std::move(controller),
                                  }));
        if (cb_result.is_error()) {
          fdf_log::error("Failed to start driver '{}', unknown request for driver {}", url,
                         moniker);
          completer.Close(ZX_ERR_UNAVAILABLE);
        }
      });
}

void Runner::handle_unknown_method(
    fidl::UnknownMethodMetadata<fuchsia_component_runner::ComponentRunner> metadata,
    fidl::UnknownMethodCompleter::Sync& completer) {
  fdf_log::warn("Unknown ComponentRunner request {}", metadata.method_ordinal);
}

zx::result<> Runner::CallCallback(const std::string& moniker,
                                  zx::result<StartedComponent> component) {
  auto it = moniker_to_owner_.find(moniker);
  if (it == moniker_to_owner_.end()) {
    return zx::error(ZX_ERR_NOT_FOUND);
  }

  std::shared_ptr owner = it->second.lock();
  if (!owner) {
    return zx::error(ZX_ERR_NOT_FOUND);
  }

  owner->OnComponentStarted(bootup_tracker_, moniker, std::move(component));
  return zx::ok();
}

}  // namespace driver_manager

File: src/devices/bin/driver_manager_rust/core/src/runner.rs
// Copyright 2026 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use crate::offer_injection::OfferInjector;
use driver_manager_types::{
    NodeOffer, OfferTransport, StartRequest, StartRequestReceiver, StartedComponent,
};
use fidl::HandleBased;
use fuchsia_component::server::{ServiceFs, ServiceObjLocal};
use futures::TryStreamExt;
use futures::channel::mpsc;
use log::{error, warn};
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;
use {
    fidl_fuchsia_component as fcomponent, fidl_fuchsia_component_decl as fdecl,
    fidl_fuchsia_component_runner as frunner, fidl_fuchsia_component_sandbox as fsandbox,
    fidl_fuchsia_process as fprocess, fuchsia_async as fasync,
};

const TOKEN_ID: u32 =
    fuchsia_runtime::HandleInfo::new(fuchsia_runtime::HandleType::User0, 0).as_raw();

#[derive(Clone)]
pub struct Runner {
    pub realm: fcomponent::RealmProxy,
    pub introspector: fcomponent::IntrospectorProxy,
    offer_injector: OfferInjector,
    pending_starts: Rc<RefCell<HashMap<String, StartRequest>>>,
    koid_to_moniker: Rc<RefCell<HashMap<zx::Koid, String>>>,
}

impl Runner {
    pub fn new(
        realm: fcomponent::RealmProxy,
        introspector: fcomponent::IntrospectorProxy,
        offer_injector: OfferInjector,
    ) -> Self {
        Self {
            realm,
            introspector,
            offer_injector,
            pending_starts: Rc::new(RefCell::new(HashMap::new())),
            koid_to_moniker: Rc::new(RefCell::new(HashMap::new())),
        }
    }

    pub fn publish<'a>(&self, fs: &mut ServiceFs<ServiceObjLocal<'a, ()>>) {
        let runner = self.clone();
        fs.dir("svc").add_fidl_service(move |stream| {
            let runner = runner.clone();
            fasync::Task::local(async move {
                if let Err(e) = runner.serve(stream).await {
                    warn!("Failed to serve ComponentRunner: {}", e);
                }
            })
            .detach();
        });
    }

    #[allow(clippy::too_many_arguments)]
    pub async fn create_driver_component(
        &self,
        moniker: &str,
        url: &str,
        collection_name: &str,
        offers: &[NodeOffer],
        dictionary_ref: Option<fsandbox::DictionaryRef>,
        skip_injected_offers: bool,
        controller: fidl::endpoints::ServerEnd<fcomponent::ControllerMarker>,
    ) -> Result<(fprocess::HandleInfo, StartRequestReceiver), zx::Status> {
        let child_decl = fdecl::Child {
            name: Some(moniker.to_string()),
            url: Some(url.to_string()),
            startup: Some(fdecl::StartupMode::Lazy),
            ..Default::default()
        };

        // Filter out the dictionary offers as they are not supported through the component
        // dynamic offers.
        let mut dynamic_offers: Vec<fdecl::Offer> = offers
            .iter()
            .filter(|offer| !matches!(offer.transport, OfferTransport::Dictionary))
            .map(|offer| {
                let fdf_offer = offer.into();
                match fdf_offer {
                    fidl_fuchsia_driver_framework::Offer::ZirconTransport(offer) => offer,
                    fidl_fuchsia_driver_framework::Offer::DriverTransport(offer) => offer,
                    _ => panic!("Unknown offer type"),
                }
            })
            .collect();

        if !skip_injected_offers {
            let extra_offers_count = self.offer_injector.extra_offers_count();
            let current_len = dynamic_offers.len();
            dynamic_offers.resize_with(current_len + extra_offers_count, || {
                fdecl::Offer::Protocol(fdecl::OfferProtocol::default())
            });
            self.offer_injector.inject(&mut dynamic_offers, current_len);
        }

        let create_child_args = fcomponent::CreateChildArgs {
            dynamic_offers: Some(dynamic_offers),
            dictionary: dictionary_ref,
            controller: Some(controller),
            ..Default::default()
        };

        let collection_ref = fdecl::CollectionRef { name: collection_name.to_string() };

        self.realm
            .create_child(&collection_ref, &child_decl, create_child_args)
            .await
            .map_err(|e| {
                error!("Failed to create child {}: {}", moniker, e);
                zx::Status::INTERNAL
            })?
            .map_err(|e| {
                error!("Failed to create child {}: {:?}", moniker, e);
                zx::Status::INTERNAL
            })?;

        let (tx, rx) = mpsc::channel(1);
        self.pending_starts.borrow_mut().insert(moniker.to_string(), tx);
        let token = zx::Event::create();
        let koid = token.basic_info()?.koid;
        self.koid_to_moniker.borrow_mut().insert(koid, moniker.to_string());
        let handle_info = fprocess::HandleInfo { handle: token.into(), id: TOKEN_ID };

        Ok((handle_info, rx))
    }

    fn complete_request(&self, moniker: &str, result: Result<StartedComponent, zx::Status>) {
        if let Some(completer) = self.pending_starts.borrow_mut().get_mut(moniker) {
            let _: Result<(), _> = completer.try_send(result).map_err(|e| {
                error!("Failed to complete request for {}: {}", moniker, e);
                zx::Status::INTERNAL
            });
        }
    }

    async fn serve(
        &self,
        mut stream: frunner::ComponentRunnerRequestStream,
    ) -> Result<(), fidl::Error> {
        while let Some(request) = stream.try_next().await? {
            match request {
                frunner::ComponentRunnerRequest::Start { start_info, controller, .. } => {
                    self.start(start_info, controller).await;
                }
                frunner::ComponentRunnerRequest::_UnknownMethod { .. } => {}
            }
        }
        Ok(())
    }

    async fn start(
        &self,
        start_info: frunner::ComponentStartInfo,
        controller: fidl::endpoints::ServerEnd<frunner::ComponentControllerMarker>,
    ) {
        let url = start_info.resolved_url.as_deref().unwrap_or("");

        let moniker = if let Some(handles) = start_info.numbered_handles.as_ref()
            && handles.len() == 1
            && handles[0].id == TOKEN_ID
        {
            let Ok(info) = handles[0].handle.basic_info() else {
                controller.close_with_epitaph(zx::Status::INVALID_ARGS).ok();
                return;
            };

            self.koid_to_moniker.borrow_mut().remove(&info.koid)
        } else {
            None
        };

        let moniker = if let Some(moniker) = moniker {
            moniker
        } else {
            // Fallback to Introspector
            let Some(component_instance) = start_info.component_instance.as_ref() else {
                error!("Failed to start driver '{}', no component instance", url);
                controller.close_with_epitaph(zx::Status::INVALID_ARGS).ok();
                return;
            };

            let Ok(token) = component_instance.duplicate_handle(zx::Rights::SAME_RIGHTS) else {
                error!("Failed to duplicate component instance for driver '{}'", url);
                controller.close_with_epitaph(zx::Status::INTERNAL).ok();
                return;
            };

            match self.introspector.get_moniker(token).await {
                Ok(Ok(moniker)) => {
                    // Moniker string is like "realm/child". We need just "child" (collection/name)?
                    // Actually Introspector returns relative moniker.
                    // C++ code does: if (split_point <= 0) error; moniker = moniker.substr(split_point + 1);
                    // Assuming moniker is "collection:name" or "name"?
                    // Wait, C++ says `moniker.find(':')`. That implies "collection:name".
                    // Let's verify what `GetMoniker` returns.
                    // The C++ code strips the collection name.
                    if let Some(pos) = moniker.find(':') {
                        moniker[pos + 1..].to_string()
                    } else {
                        error!(
                            "Moniker '{}' for driver '{}' does not contain collection",
                            moniker, url
                        );
                        controller.close_with_epitaph(zx::Status::INVALID_ARGS).ok();
                        return;
                    }
                }
                Ok(Err(e)) => {
                    error!("Failed to get moniker for driver '{}': {:?}", url, e);
                    controller.close_with_epitaph(zx::Status::INTERNAL).ok();
                    return;
                }
                Err(e) => {
                    error!("Failed to call GetMoniker for driver '{}': {}", url, e);
                    controller.close_with_epitaph(zx::Status::INTERNAL).ok();
                    return;
                }
            }
        };

        if self.pending_starts.borrow().get(&moniker).is_none() {
            error!("Failed to start driver '{}', unknown request for driver {}", url, moniker);
            controller.close_with_epitaph(zx::Status::UNAVAILABLE).ok();
            return;
        }

        self.complete_request(&moniker, Ok(StartedComponent { info: start_info, controller }));
    }
}
